code/datums/storage/storage.dm

```diff
@@ -0,0 +1,1029 @@
+
+	if(new_location)
+		if(ismob(new_location))
+			item.layer = ABOVE_HUD_LAYER
+			item.plane = ABOVE_HUD_PLANE
+			item.pickup(new_location)
+		else
+			item.layer = initial(item.layer)
+			item.plane = initial(item.plane)
+		item.forceMove(new_location)
+	else
+		item.moveToNullspace()
+
+	orient2hud()
+
+	for(var/i in can_see_content())
+		var/mob/M = i
+		show_to(M)
+
+	if(!QDELETED(item))
+		item.on_exit_storage(src)
+		item.mouse_opacity = initial(item.mouse_opacity)
+
+	for(var/limited_type in storage_type_limits_max)
+		if(istype(item, limited_type))
+			storage_type_limits_max[limited_type] += 1
+
+	parent.update_icon()
+
+	return TRUE
+
+///Refills the storage from the refill_types item
+/datum/storage/proc/do_refill(obj/item/storage/refiller, mob/user)
+	if(!length(refiller.contents))
+		user.balloon_alert(user, "[refiller] is empty.")
+		return
+
+	if(!can_be_inserted(refiller.contents[1]))
+		user.balloon_alert(user, "[parent.name] is full.")
+		return
+
+	user.balloon_alert(user, "Refilling.")
+
+	if(!do_after(user, 15, NONE, src, BUSY_ICON_GENERIC))
+		return
+
+	playsound(user.loc, refill_sound, 15, 1, 6)
+	for(var/obj/item/IM in refiller)
+		if(!can_be_inserted(refiller.contents[1]))
+			return
+
+		remove_from_storage(IM)
+		handle_item_insertion(IM, TRUE, user)
+
+/datum/storage/proc/quick_empty()
+
+	if((!ishuman(usr) && parent.loc != usr) || usr.restrained())
+		return
+
+	var/turf/T = get_turf(src)
+	hide_from(usr)
+	for(var/obj/item/item in parent.contents)
+		remove_from_storage(item, T, usr)
+
+///Delete everything that's inside the storage
+/datum/storage/proc/delete_contents()
+	for(var/obj/item/item AS in parent.contents)
+		item.on_exit_storage(src)
+		qdel(item)
+
+///Returns the storage depth of an atom. This is the number of storage items the atom is contained in before reaching toplevel (the area). Returns -1 if the atom was not found on container.
+/datum/storage/proc/storage_depth(atom/container)
+	var/depth = 0
+	var/atom/cur_atom = src
+
+	while (cur_atom && !(cur_atom in container.contents))
+		if (isarea(cur_atom))
+			return -1
+		if (istype(cur_atom.loc, /obj/item/storage))
+			depth++
+		cur_atom = cur_atom.loc
+
+	if (!cur_atom)
+		return -1	//inside something with a null loc.
+
+	return depth
+
+///Like storage depth, but returns the depth to the nearest turf. Returns -1 if no top level turf (a loc was null somewhere, or a non-turf atom's loc was an area somehow).
+/datum/storage/proc/storage_depth_turf()
+	var/depth = 0
+	var/atom/cur_atom = src
+
+	while (cur_atom && !isturf(cur_atom))
+		if (isarea(cur_atom))
+			return -1
+		if (istype(cur_atom.loc, /obj/item/storage))
+			depth++
+		cur_atom = cur_atom.loc
+
+	if (!cur_atom)
+		return -1	//inside something with a null loc.
+
+	return depth
+
+//Equips an item from our storage, returns signal COMSIG_QUICK_EQUIP_HANDLED to prevent standard quick equip behaviour
+/datum/storage/proc/on_quick_equip_request(datum/source, mob/user)
```

Comment:
```suggestion
/// Equips an item from our storage, returns signal COMSIG_QUICK_EQUIP_HANDLED to prevent standard quick equip behaviour
/datum/storage/proc/on_quick_equip_request(datum/source, mob/user)
```